File: ios-native/Spottr/Analyzers/Geometry.swift
import CoreGraphics
import Foundation

enum Geometry {
    /// Angle ABC in degrees (joint at B). Returns NaN if any point is missing.
    static func angle(_ a: Keypoint?, _ b: Keypoint?, _ c: Keypoint?) -> Double {
        guard let a = a, let b = b, let c = c else { return .nan }
        let v1 = CGVector(dx: a.x - b.x, dy: a.y - b.y)
        let v2 = CGVector(dx: c.x - b.x, dy: c.y - b.y)
        let dot = v1.dx * v2.dx + v1.dy * v2.dy
        let m1 = (v1.dx * v1.dx + v1.dy * v1.dy).squareRoot()
        let m2 = (v2.dx * v2.dx + v2.dy * v2.dy).squareRoot()
        guard m1 > 0, m2 > 0 else { return .nan }
        let cos = max(-1.0, min(1.0, Double(dot / (m1 * m2))))
        return acos(cos) * 180.0 / .pi
    }

    static func midpoint(_ a: Keypoint?, _ b: Keypoint?) -> CGPoint? {
        guard let a = a, let b = b else { return nil }
        return CGPoint(x: (a.x + b.x) / 2, y: (a.y + b.y) / 2)
    }

    static func avgScore(_ kps: Keypoint?...) -> Double {
        let valid = kps.compactMap { $0 }
        guard !valid.isEmpty else { return .nan }
        return valid.map(\.score).reduce(0, +) / Double(valid.count)
    }

    static func avgFinite(_ xs: Double...) -> Double {
        let ok = xs.filter { $0.isFinite }
        guard !ok.isEmpty else { return .nan }
        return ok.reduce(0, +) / Double(ok.count)
    }

    static func scoreFromIssues(_ issues: [FormIssue]) -> Double {
        var score = 1.0
        for i in issues {
            switch i.severity {
            case .severe:   score -= 0.35
            case .moderate: score -= 0.20
            case .minor:    score -= 0.08
            }
        }
        return max(0, min(1, score))
    }
}

File: app/src/lib/exercises/geometry.ts
import type { Keypoint } from './types';

/** Angle ABC in degrees (joint at B). Returns NaN if any point is missing. */
export function angleDeg(a?: Keypoint, b?: Keypoint, c?: Keypoint): number {
  if (!a || !b || !c) return NaN;
  const v1x = a.x - b.x;
  const v1y = a.y - b.y;
  const v2x = c.x - b.x;
  const v2y = c.y - b.y;
  const dot = v1x * v2x + v1y * v2y;
  const m1 = Math.hypot(v1x, v1y);
  const m2 = Math.hypot(v2x, v2y);
  if (m1 === 0 || m2 === 0) return NaN;
  const cos = Math.max(-1, Math.min(1, dot / (m1 * m2)));
  return (Math.acos(cos) * 180) / Math.PI;
}

export function midpoint(a?: Keypoint, b?: Keypoint): { x: number; y: number } | null {
  if (!a || !b) return null;
  return { x: (a.x + b.x) / 2, y: (a.y + b.y) / 2 };
}

/** Average confidence of provided keypoints (NaN if none). */
export function avgScore(...kps: Array<Keypoint | undefined>): number {
  const xs = kps.filter((k): k is Keypoint => Boolean(k));
  if (xs.length === 0) return NaN;
  return xs.reduce((s, k) => s + k.score, 0) / xs.length;
}
